cmd/qat_plugin/qat_plugin.go

```diff
@@ -0,0 +1,358 @@
+	devicePCIAddrBytes := []byte(devicePCIAddr)
+	vfdevID, err := getDeviceID(devicePCIAddr)
+	if err != nil {
+		glog.Error(err)
+		fmt.Printf("Cannot obtain the Device ID for this device")
+		return fmt.Errorf("Cannot obtain the Device ID for this device: %v\n", err)
+	}
+	vfBindStr := concatStr(vendorPrefix, vfdevID)
+	vfBindStrBytes := []byte(vfBindStr)
+	err = ioutil.WriteFile(unbindKernelDevicePath, devicePCIAddrBytes, 0644)
+	if err != nil {
+		glog.Error(err)
+		fmt.Printf("Unbinding from the kernel driver failed\n")
+		return fmt.Errorf("Unbinding from the kernel driver failed: %v\n", err)
+
+	}
+	// Unbinding from the kernel driver DONE
+	err = ioutil.WriteFile(bindDevicePath, vfBindStrBytes, 0644)
+	if err != nil {
+		fmt.Printf("Binding to the dpdk driver failed\n")
+		return fmt.Errorf("Binding to the dpdk driver failed: %v\n", err)
+	}
+	//Binding to the the dpdk driver DONE\n
+	return nil
+}
+func isValidKerneDriver(kernelvfDriver string) error {
+	switch kernelvfDriver {
+	case "dh895xccvf", "c6xxvf", "c3xxxvf", "d15xxvf":
+		return nil
+	default:
+		return fmt.Errorf("Please enter the kernel-vf-drivers flag correctly")
+	}
+}
+func isValidDpdkDeviceDriver(dpdkDriver string) error {
+	switch dpdkDriver {
+	case "igb_uio", "vfio-pci":
+		return nil
+	default:
+		return fmt.Errorf("Please enter the dpdk-driver flag correctly")
+	}
+}
+
+// Discovers all QAT devices available on the local node by querying PCI bus using lspci.
+func (dm *deviceManager) discoverQATs() (bool, error) {
+
+	var found bool = false
+	fmt.Println("Discovered Devices below:")
+	kernelvfDrivers := strings.Split(*kernelVfDrivers, ",")
+
+	for _, kernelvfDriver := range kernelvfDrivers {
+
+		err := isValidKerneDriver(kernelvfDriver)
+		if err != nil {
+			return found, fmt.Errorf("Error in user input for kernel VF Driver: %v", err)
+		}
+
+		err = isValidDpdkDeviceDriver(*dpdkDriver)
+		if err != nil {
+			return found, fmt.Errorf("Error in user input for DPDK Device Driver: %v", err)
+		}
+
+		files, err := ioutil.ReadDir(concatStr(pciDriverDir, kernelvfDriver))
+		if err != nil {
+			fmt.Errorf("Can't read sysfs for kernel vf driver %v: %v", kernelvfDriver, err)
+			continue
+		}
+
+		for n, file := range files {
+			if strings.HasPrefix(file.Name(), "0000:") {
+				vfpciaddr := strings.TrimPrefix(file.Name(), "0000:")
+				max, err := strconv.Atoi(*maxNumdevice)
+				if err != nil {
+					return found, fmt.Errorf("Error in getting maximum number of devices: %v", err)
+				}
+
+				if n < max {
+					//vfpciaddr := strings.Fields(vf)[0]
+
+					err = BindDevice(*dpdkDriver, vfpciaddr)
+					if err != nil {
+						return found, fmt.Errorf("Error in binding the device to the dpdk driver")
+					}
+					devNodes, err := getDpdkDeviceNames(vfpciaddr)
+					if err != nil {
+						return found, fmt.Errorf("Error in obtaining the device name ")
+					}
+					devMountPoints, err := getDpdkMountPaths(vfpciaddr)
```

Comment:
I first noticed in the YAML file, but here I see mount paths in `/sys` are added to the response. I thought `/sys` is always available inside containers.

Reply:
The default sysfs which appears in the container is read only. But in our case we are writing to the sysfs when the binding unbinding takes places. If we don't mount /sys/bus/pci writing file /sys/bus/pci/devices/<device PCI address>/driver/unbind fails. For security reasons I don't mount in the whole sysfs.